Stack/Opc.Ua.Core/Stack/Tcp/TcpTransportListener.cs

```diff
@@ -526,63 +526,72 @@ private void OnAccept(object sender, SocketAsyncEventArgs e)
+                            try
                             {
-                                // create the channel to manage incoming reverse connections.
-                                channel = new TcpReverseConnectChannel(
-                                    m_listenerId,
-                                    this,
-                                    m_bufferManager,
-                                    m_quotas,
-                                    m_descriptions);
+                                if (m_reverseConnectListener)
+                                {
+                                    // create the channel to manage incoming reverse connections.
+                                    channel = new TcpReverseConnectChannel(
+                                        m_listenerId,
+                                        this,
+                                        m_bufferManager,
+                                        m_quotas,
+                                        m_descriptions);
+                                }
+                                else
+                                {
+                                    // create the channel to manage incoming connections.
+                                    channel = new TcpServerChannel(
+                                        m_listenerId,
+                                        this,
+                                        m_bufferManager,
+                                        m_quotas,
+                                        m_serverCertificate,
+                                        m_serverCertificateChain,
+                                        m_descriptions);
+                                }
+
+                                if (m_callback != null)
+                                {
+                                    channel.SetRequestReceivedCallback(new TcpChannelRequestEventHandler(OnRequestReceived));
+                                    channel.SetReportOpenSecureChannelAuditCallback(new ReportAuditOpenSecureChannelEventHandler(OnReportAuditOpenSecureChannelEvent));
+                                    channel.SetReportCloseSecureChannelAuditCallback(new ReportAuditCloseSecureChannelEventHandler(OnReportAuditCloseSecureChannelEvent));
+                                    channel.SetReportCertificateAuditCallback(new ReportAuditCertificateEventHandler(OnReportAuditCertificateEvent));
+                                }
+
+                                uint channelId;
+                                do
+                                {
+                                    // get channel id
+                                    channelId = GetNextChannelId();
+
+                                    // save the channel for shutdown and reconnects.
+                                    // retry to get a channel id if it is already in use.
+                                } while (!channels.TryAdd(channelId, channel));
```

Comment:
you add into your local channels copy not the field this seems not right

Reply:
no thats correct and intended, its not a local copy, just a local reference which is already null checked, thats the idea.